### PR TITLE
Add and use `DeepReveal` trait

### PR DESCRIPTION
**Pull request by MingweiSamuel**

### CODE REVIEW

File: lattices/src/collections.rs [hunk 1]
```diff
@@ -10,6 +11,47 @@ use cc_traits::{
     KeyedRef, Len, MapIter, MapIterMut, SimpleKeyedRef,
 };
 
+/// Trait for transforming the values of a map without changing the overal type of the data structure.
```
Comment (jhellerstein) [inline lattices/src/collections.rs:14]: nit: typo "overal"

Comment (jhellerstein) [reply lattices/src/collections.rs:14]: For all the below, are there cases where in-place mutation would be faster and more desirable than map? Should we support that in addition?

**Review by jhellerstein — APPROVED** [review]
LGTM for now. Would it be more elegant to do something like a general-purpose visitor pattern, with reveal as just one of the things one might do at each level? Thinking about possible applications of that -- one might be pretty-printing or diagram generation.

**Comment by MingweiSamuel** [comment]
If we need it, it may be possible to do, later

**Pull request merged**